Dev

KHudali17 opened the pull request.

Commit:
removed unnecessary condition

Commit:
Implemented adding-a-product functionality.

Implemented the Invoke class, the AddProduct method of class Inventory, and added AddProductCommand class

Commit:
Added search-by-name functionality

Created a RetrieveByNameCommand class to handle user prompt and call the search method

Commit:
Added view-all functionality

Created a RetrieveAllCommand class to retrieve and print the product list

Commit:
Added delete-product functionality

Created a RemoveProductCommand class to delete a product from inventory.

Commit:
Added update-product functionality

Create an UpdateProductCommand class to update a product from inventory.

MazenAmria commented on Inventory.cs:
```diff
@@ -3,34 +3,41 @@
 {
     public class Inventory : IInventory
     {
+        private List<Product> products;
```
> Use camel casing ("camelCasing") when naming private or internal fields, and prefix them with _.

https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions#camel-case

MazenAmria commented on Commands.cs:
```diff
@@ -0,0 +1,190 @@
+        {
+            this.inventory = inventory;
+        }
+        public void Execute()
```
> Add at least one blank line between method definitions and property definitions

https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/coding-conventions#layout-conventions

MazenAmria commented on Commands.cs:
```diff
@@ -0,0 +1,190 @@
+            //Print list if not empty
+            if(products.Count == 0)
+            {
+                Console.WriteLine("Inventory is Empty.");
```
it would be better if you returned the execution to avoid the two-levels nesting
```CSharp
if (isEmpty)
{
    // ...
    return;
}

foreach(...)
{
    // ...
}
```

MazenAmria reviewed (changes requested):
1. split the commands into separate files but within the same folder and namespace
2. try throwing exceptions for `product not found` and `inventory is empty`, but I'm not sure if this is going to complicate things as we have a common interface for the Command and the invoker is acting on that interface, so it should be aware of the exceptions. @SwishBoy what do you think

Commit:
Fixed code-style and code-layout issues

Corrected private and readonly attribute names to begin with underscore; Added a space between property definitions and method definitions. Moved each of the classes in Commands.cs into its own file.

kshawish commented on AddProductsCommand.cs:
```diff
@@ -0,0 +1,44 @@
+using System.Text;
+using System.Threading.Tasks;
+
+namespace SimpleInv
```
let's avoid abbreviations especially when the term is not excessively long. 
we typically reserve abbreviations for terms composed of multiple words and are commonly used within the project domain.

kshawish commented on Inventory.cs:
```diff
@@ -3,34 +3,42 @@
+            return _products;
         }
 
         public void UpdateProduct(Product product, string newName, decimal newPrice, decimal newQuantity)
```
Just worndering about the update method here.
Usually I'd expect to pass and object that represent the updated Product to the Inventory update method which will reflect that to the In-Memery/DB product. and not passing each property we wanna update to the update method.

kshawish commented on Inventory.cs:
```diff
@@ -3,34 +3,42 @@
+        public Inventory() {
+            _products = new List<Product>();
+        }
         public void AddProduct(Product product)
```
I don't see any validation rules in the Inventory? what if we tried to add a product with negative price ?

kshawish commented on Inventory.cs:
```diff
@@ -3,34 +3,42 @@
         public Product Retrieve(string productName)
         {
-            throw new NotImplementedException();
+            return _products.Find(x => x.Name.Equals(productName, StringComparison.OrdinalIgnoreCase));
```
what if the product is not found will you depend on find throw exception? or do you think it's better to catch it and throw a custom exception?
